File: ts/project/.sdk/tm/js/src/feature/cache/CacheFeature.js


const { BaseFeature } = require('../base/BaseFeature')


// Response caching for safe (read) requests. Wraps the active transport and
// serves a fresh cached snapshot instead of hitting the network when the
// same method+URL was fetched within `ttl` ms. Only successful responses to
// cacheable methods (default: GET) are stored, keyed by method+URL. The
// cache is bounded (`max` entries, oldest evicted) and every hit/miss is
// recorded on `client._cache` for inspection. One-shot response bodies are
// normalised on capture so both the current caller and later hits can read
// the JSON body repeatedly.
class CacheFeature extends BaseFeature {
  version = '0.0.1'
  name = 'cache'
  active = true

  _client
  _options = {}
  _store = new Map()


  init(ctx, options) {
    this._client = ctx.client
    this._options = options || {}
    this.active = options.active

    if (!this.active) {
      return
    }

    this._store = new Map()

    const self = this
    const utility = ctx.utility
    const inner = utility.fetcher

    utility.fetcher = async function (ctx2, url, fetchdef) {
      return self._through(ctx2, url, fetchdef, inner)
    }
  }


  async _through(ctx, url, fetchdef, inner) {
    const method = ((fetchdef && fetchdef.method) || 'GET').toUpperCase()
    const methods = this._options.methods || ['GET']

    if (methods.indexOf(method) < 0) {
      return inner(ctx, url, fetchdef)
    }

    const key = method + ' ' + url
    const now = this._now()
    const hit = this._store.get(key)

    if (null != hit && hit.expiry > now) {
      this._track('hit')
      return this._replay(hit.snapshot)
    }

    const res = await inner(ctx, url, fetchdef)

    if (this._cacheable(res)) {
      const snapshot = await this._snapshot(res)
      const ttl = null == this._options.ttl ? 5000 : this._options.ttl
      this._evict()
      this._store.set(key, { expiry: now + ttl, snapshot })
      this._track('miss')
      return this._replay(snapshot)
    }

    this._track('bypass')
    return res
  }


  _cacheable(res) {
    if (null == res || res instanceof Error) {
      return false
    }
    const status = res.status
    return 'number' === typeof status && status >= 200 && status < 300
  }


  async _snapshot(res) {
    let data = undefined
    if ('function' === typeof res.json) {
      try { data = await res.json() } catch (_e) { data = undefined }
    }
    const headers = {}
    if (res.headers && 'function' === typeof res.headers.forEach) {
      res.headers.forEach((v, k) => headers[k] = v)
    }
    return { status: res.status, statusText: res.statusText, data, headers }
  }


  _replay(snapshot) {
    const headers = snapshot.headers || {}
    return {
      status: snapshot.status,
      statusText: snapshot.statusText,
      body: 'not-used',
      json: async () => snapshot.data,
      headers: {
        get(key) { return headers[String(key).toLowerCase()] },
        forEach(callback) {
          Object.keys(headers).forEach((k) => callback(headers[k], k, this))
        },
      },
    }
  }


  _evict() {
    const max = null == this._options.max ? 256 : this._options.max
    while (this._store.size >= max) {
      const oldest = this._store.keys().next().value
      if (null == oldest) {
        break
      }
      this._store.delete(oldest)
    }
  }


  _now() {
    const now = this._options.now
    if ('function' === typeof now) {
      return now()
    }
    return Date.now()
  }


  _track(kind) {
    const client = this._client
    if (null == client._cache) {
      client._cache = { hit: 0, miss: 0, bypass: 0 }
    }
    client._cache[kind]++
  }
}


module.exports = {
  CacheFeature
}

File: ts/project/.sdk/tm/swift/Sources/ProjectNameSDK/feature/CacheFeature.swift
// Response caching for safe (read) requests. Wraps the active transport and
// serves a fresh cached snapshot instead of hitting the network when the same
// method+URL was fetched within `ttl` ms (default: 5000). Only successful
// (2xx) responses to cacheable methods (default: GET) are stored, keyed by
// method+URL. The cache is bounded (`max` entries, default 256, oldest evicted
// first) and every hit/miss/bypass is counted. Bodies are snapshotted on
// capture so both the current caller and later hits can re-read the JSON body
// repeatedly.

import Foundation

public final class CacheFeature: BaseFeature {
  private var client: ProjectNameSDK?
  private var options: VMap?
  private var store: [String: CacheEntry] = [:]
  private var order: [String] = []

  // Activity tracking (mirrors the ts client._cache record).
  public var hit = 0
  public var miss = 0
  public var bypass = 0

  private final class CacheEntry {
    var expiry: Int64 = 0
    var snapshot: CacheSnapshot = CacheSnapshot()
  }

  private final class CacheSnapshot {
    var status: Int = 0
    var statusText: String = ""
    var data: Value = .noval
    var headers: VMap = VMap()
  }

  public override init() {
    super.init()
    version = "0.0.1"
    name = "cache"
    active = true
  }

  public override func initFeature(_ ctx: Context, _ options: VMap) {
    client = ctx.client
    self.options = options
    active = foptBool(options, "active", false)

    if !active {
      return
    }

    store = [:]
    order = []

    let inner = ctx.utility!.fetcher!

    ctx.utility!.fetcher = { ctx2, url, fetchdef in
      try self.through(ctx2, url, fetchdef, inner)
    }
  }

  private func through(_ ctx: Context, _ url: String, _ fetchdef: VMap,
                       _ inner: FetcherFunc) throws -> Value {
    var method = "GET"
    if let m = fetchdef.entries["method"]?.asString, m != "" {
      method = m.uppercased()
    }

    let methods = foptStrList(options, "methods") ?? ["GET"]
    let cacheable = methods.contains { $0.uppercased() == method }
    if !cacheable {
      return try inner(ctx, url, fetchdef)
    }

    let key = method + " " + url
    let now = foptNow(options)()

    if let entry = store[key], entry.expiry > now {
      hit += 1
      return CacheFeature.replay(entry.snapshot)
    }

    var res: Value = .noval
    var err: Error? = nil
    do {
      res = try inner(ctx, url, fetchdef)
    } catch {
      err = error
    }

    if err == nil && CacheFeature.storable(res) {
      let snap = CacheFeature.snapshot(res)
      let ttl = foptInt(options, "ttl", 5000)
      evict()
      let entry = CacheEntry()
      entry.expiry = now + Int64(ttl)
      entry.snapshot = snap
      store[key] = entry
      order.append(key)
      miss += 1
      return CacheFeature.replay(snap)
    }

    bypass += 1
    if let err = err {
      throw err
    }
    return res
  }

  private static func storable(_ res: Value) -> Bool {
    let (status, has) = fresStatus(res)
    return has && status >= 200 && status < 300
  }

  private static func snapshot(_ res: Value) -> CacheSnapshot {
    let snap = CacheSnapshot()

    let (status, has) = fresStatus(res)
    if has {
      snap.status = status
    }
    if let rm = res.asMap {
      if let st = rm.entries["statusText"]?.asString {
        snap.statusText = st
      }
      if let f = rm.entries["json"]?.asNative as? NativeCall0 {
        snap.data = f()
      }
      if let headers = rm.entries["headers"]?.asMap {
        for (k, v) in headers.entries {
          snap.headers.entries[k.lowercased()] = v
        }
      }
    }

    return snap
  }

  // Replay builds a fresh transport-shaped response so the body stays
  // re-readable for every consumer.
  private static func replay(_ snap: CacheSnapshot) -> Value {
    let data = snap.data
    let headers = VMap()
    for (k, v) in snap.headers.entries {
      headers.entries[k] = v
    }
    let r = VMap()
    r.entries["status"] = .int(Int64(snap.status))
    r.entries["statusText"] = .string(snap.statusText)
    r.entries["body"] = .string("not-used")
    r.entries["json"] = .nat({ () -> Value in data } as NativeCall0)
    r.entries["headers"] = .map(headers)
    return .map(r)
  }

  // Evict drops oldest entries (FIFO) until the store is under `max`.
  private func evict() {
    let max = foptInt(options, "max", 256)
    while store.count >= max && order.count > 0 {
      let oldest = order.removeFirst()
      store.removeValue(forKey: oldest)
    }
  }
}
